# SHARK ATTACK - DATA CLEANING

## Goal
The objective of this script is to clean the dataset of global shark attack, downloaded from [kaggle](https://www.kaggle.com/teajay/global-shark-attacks), last updated on 2018-07-04, for further analysis on relevant and reliable information.

The cleaning should be done to answer questions such as:
- What countrys have more shark attacks? Which ones have higher fatality rates?
- Is it possible to determine a typical profile of the victim? If it is, what is this profile? (Age, gender)
- What kind of activity is more vulnarable to shark attacks? What are most fatals?
- Is there a trend of raise or decrease in the number cases over the years?

In [1]:
import pandas as pd
import numpy as np
import re

In [2]:
!head data/attacks.csv

Case Number,Date,Year,Type,Country,Area,Location,Activity,Name,Sex ,Age,Injury,Fatal (Y/N),Time,Species ,Investigator or Source,pdf,href formula,href,Case Number,Case Number,original order,,
2018.06.25,25-Jun-2018,2018,Boating,USA,California,"Oceanside, San Diego County",Paddling,Julie Wolfe,F,57,"No injury to occupant, outrigger canoe and paddle damaged",N,18h00,White shark,"R. Collier, GSAF",2018.06.25-Wolfe.pdf,http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf,http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf,2018.06.25,2018.06.25,6303,,
2018.06.18,18-Jun-2018,2018,Unprovoked,USA,Georgia,"St. Simon Island, Glynn County",Standing,Adyson McNeely ,F,11,Minor injury to left thigh,N,14h00  -15h00,,"K.McMurray, TrackingSharks.com",2018.06.18-McNeely.pdf,http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf,http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf,2018.06.18,2018.06.18,6302,,
2018.0

In [3]:
#read the .csv data file
#encoding is 'latin-1', if leave encoding at default value, it gives error
df_attacks = pd.read_csv('data/attacks.csv', encoding='latin-1')

In [4]:
pd.set_option('display.max_columns', None) #display all columns
df_attacks

Case Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
25718         NaN          NaN     NaN         NaN        NaN   
25719         NaN          NaN     NaN         NaN        NaN   
25720         NaN          NaN     NaN         NaN        NaN   
25721         NaN          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free diving   
...                ...                             ...          ...   
25718              NaN                             NaN          NaN   
25719              NaN                             NaN          NaN   
25720              NaN                             NaN          NaN   
25721              NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                  Name Sex   Age  \
0          Julie Wolfe    F   57   
1      Adyson McNeely     F   11   
2          John Denges    M   48   
3                 male    M  NaN   
4       Gustavo Ramos     M  NaN   
...                ...  ...  ...   
25718              NaN  NaN  NaN   
25719              NaN  NaN  NaN   
25720              NaN  NaN  NaN   
25721              NaN  NaN  NaN   
25722              NaN  NaN  NaN   

                                                  Injury Fatal (Y/N)  \
0      No injury to occupant, outrigger canoe and pad...           N   
1                             Minor injury to left thigh           N   
2           Injury to left lower leg from surfboard skeg           N   
3                              Minor injury to lower leg           N   
4      Lacerations to leg & hand shark PROVOKED INCIDENT           N   
...                                                  ...         ...   
25718                                                NaN         NaN   
25719                                                NaN         NaN   
25720                                                NaN         NaN   
25721                                                NaN         NaN   
25722                                                NaN         NaN   

                Time         Species           Investigator or Source  \
0              18h00      White shark                R. Collier, GSAF   
1      14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2              07h45              NaN  K.McMurray, TrackingSharks.com   
3                NaN        2 m shark                  B. Myatt, GSAF   
4                NaN  Tiger shark, 3m                       A .Kipper   
...              ...              ...                             ...   
25718            NaN              NaN                             NaN   
25719            NaN              NaN                             NaN   
25720            NaN              NaN                             NaN   
25721            NaN              NaN                             NaN   
25722            NaN              NaN                             NaN   

                            pdf  \
0          2018.06.25-Wolfe.pdf   
1        2018.06.18-McNeely.pdf   
2         2018.06.09-Denges.pdf   
3      2018.06.08-Arrawarra.pdf   
4          2018.06.04-

In [5]:
df_attacks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

In [6]:
df_attacks['Unnamed: 22'].value_counts()

stopped here    1
Name: Unnamed: 22, dtype: int64

In [7]:
df_attacks['Unnamed: 23'].value_counts()

Teramo             1
change filename    1
Name: Unnamed: 23, dtype: int64

**Note 1: Columns 22 and 23 are useless, no valid information because it has only one or two non-null values.**

In [8]:
df_attacks.describe().T

count         mean          std  min     25%     50%     75%  \
Year            6300.0  1927.272381   281.116308  0.0  1942.0  1977.0  2005.0   
original order  6309.0  3155.999683  1821.396206  2.0  1579.0  3156.0  4733.0   

                   max  
Year            2018.0  
original order  6310.0

There is no much useful information with this `describe()`. Actually, the best information is that the most recent attack registered in this dataset ocurred in 2018.

The `original order` also does not add anything meaningful.

In [9]:
#checking how many duplicateds
df_attacks.duplicated().sum()

19411

Lots of dupes: **19411**

---
## Start cleaning

In [10]:
#droping duplicates
df_attacks.drop_duplicates(inplace=True)

#droping useless columns for exploratory analysis
columns_to_drop = ['Case Number', 'Name', 'Investigator or Source', 'pdf', 'href formula','href',\
                   'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22', 'Unnamed: 23']
df_attacks.drop(columns=columns_to_drop, inplace=True)

In [11]:
#cleaning to analyse only the last 50 years of records
df_attacks.query('1969 <= Year <= 2018', inplace=True)
df_attacks

Date    Year        Type    Country             Area  \
0     25-Jun-2018  2018.0     Boating        USA       California   
1     18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2     09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3     08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4     04-Jun-2018  2018.0    Provoked     MEXICO           Colima   
...           ...     ...         ...        ...              ...   
3441     Feb-1969  1969.0     Invalid  AUSTRALIA       Queensland   
3442  27-Jan-1969  1969.0  Unprovoked  AUSTRALIA  New South Wales   
3443     Jan-1969  1969.0  Unprovoked  AUSTRALIA         Victoria   
3444  Winter 1969  1969.0    Provoked    BAHAMAS        Eleuthera   
3888  16-Jul-1961  1971.0  Unprovoked     TURKEY         Anatolia   

                            Location         Activity Sex   Age  \
0        Oceanside, San Diego County         Paddling    F   57   
1     St. Simon Island, Glynn County         Standing    F   11   
2                       Habush, Oahu          Surfing    M   48   
3                 Arrawarra Headland          Surfing    M  NaN   
4                           La Ticla      Free diving    M  NaN   
...                              ...              ...  ...  ...   
3441                        Cooktown  Hard hat diving    M   52   
3442                   Beecroft Head       Freediving    M   21   
3443                  Port MacDonnel              NaN    M  NaN   
3444                             NaN     Sight-seeing  NaN  NaN   
3888         ?nciralti Beach, ?zmir          Swimming    M   16   

                                                 Injury Fatal (Y/N)  \
0     No injury to occupant, outrigger canoe and pad...           N   
1                            Minor injury to left thigh           N   
2          Injury to left lower leg from surfboard skeg           N   
3                             Minor injury to lower leg           N   
4     Lacerations to leg & hand shark PROVOKED INCIDENT           N   
...                                                 ...         ...   
3441                                         No details         NaN   
3442       Abrasions and lacerations to lower right leg           N   
3443                                      Minor injury            N   
3444  No injury to occupants, shark grabbed prop, Le...           N   
3888                                   Left leg injured           N   

               Time               Species   
0             18h00            White shark  
1     14h00  -15h00                    NaN  
2             07h45                    NaN  
3               NaN              2 m shark  
4               NaN        Tiger shark, 3m  
...             ...                    ...  
3441            NaN  Questionable incident  
3442          07h30                    NaN  
3443            NaN           Carpet shark  
3444            NaN              10' shark  
3888            NaN                    NaN  

[3445 rows x 13 columns]

---
### Cleaning Type column

In [12]:
df_attacks['Type'].value_counts()

Unprovoked      2697
Invalid          273
Provoked         257
Boating           92
Boat              69
Sea Disaster      52
Questionable       2
Boatomg            1
Name: Type, dtype: int64

`Boating`, `Boat` and `Boatomg` should be only one thing.

In [13]:
df_attacks['Type'].isna().sum() #checking null/Na values to treat them first

2

In [14]:
df_attacks['Type'].apply(type).value_counts()

<class 'str'>      3443
<class 'float'>       2
Name: Type, dtype: int64

In [15]:
#treating any NaN or null values type to become a string 'Unknown'
df_attacks['Type'] = df_attacks['Type'].apply(lambda x: x if type(x) == str else 'Unknown')
df_attacks['Type'].apply(type).value_counts()

<class 'str'>    3445
Name: Type, dtype: int64

In [16]:
#now that all values from this column is string type, we can treat the Boat/Boating/Boatomg
df_attacks['Type'] = df_attacks['Type'].apply(lambda x: 'Boating' if 'Boat' in x else x)
df_attacks['Type'].value_counts()

Unprovoked      2697
Invalid          273
Provoked         257
Boating          162
Sea Disaster      52
Questionable       2
Unknown            2
Name: Type, dtype: int64

Further analysis will be need to understand if `Questionable`, `Invalid` and `Unknown` types are to be used.

---

### Cleaning Country column
To be done

In [17]:
df_attacks['Country'].value_counts()

USA             1614
AUSTRALIA        579
SOUTH AFRICA     363
BRAZIL           103
BAHAMAS           82
                ... 
Seychelles         1
LIBERIA            1
MEXICO             1
HONDURAS           1
TURKEY             1
Name: Country, Length: 120, dtype: int64

Certainly `MEXICO` has more than 1 register.

In [18]:
len(df_attacks[df_attacks['Country'] == 'MEXICO '])

1

In [19]:
len(df_attacks[df_attacks['Country'] == 'MEXICO'])

40

---

### Cleaning Activity column

In [20]:
df_attacks['Activity'].value_counts()

Surfing                                           943
Swimming                                          437
Spearfishing                                      216
Fishing                                           170
Wading                                            112
                                                 ... 
Swimming with pod of dolphins                       1
Walking, carrying surfboard & stepped on shark      1
Swimming, wearing black wetsuit & swim fins         1
Standing, stepped on shark                          1
Sight-seeing                                        1
Name: Activity, Length: 595, dtype: int64

Too many `Activity` types. This must be simplified if it is to be used to answer the business question.

In [22]:
#testing the idea of how to simplify texts containing the keywords for the Activity type
text = pd.Series(['Swimming with pod of dolphins', 'Swimming, wearing black wetsuit & swim fins'])
text

0                  Swimming with pod of dolphins
1    Swimming, wearing black wetsuit & swim fins
dtype: object

In [23]:
pattern = '[Ss]wimming'
text.apply(lambda x: re.findall(pattern, x)[0])

0    Swimming
1    Swimming
dtype: object

In [24]:
#checking how many rows in the column 'Activity' is not a string
df_attacks['Activity'].apply(type).value_counts()

<class 'str'>      3229
<class 'float'>     216
Name: Activity, dtype: int64

In [25]:
#turning float types 'Activity' into 'Unknown'
df_attacks['Activity'] = df_attacks['Activity'].apply(lambda x: x if type(x) == str else 'Unknown')
df_attacks['Activity'].apply(type).value_counts()

<class 'str'>    3445
Name: Activity, dtype: int64

In [26]:
df_attacks['Activity'].value_counts()

Surfing                                           943
Swimming                                          437
Unknown                                           219
Spearfishing                                      216
Fishing                                           170
                                                 ... 
Swimming with pod of dolphins                       1
Walking, carrying surfboard & stepped on shark      1
Swimming, wearing black wetsuit & swim fins         1
Standing, stepped on shark                          1
Sight-seeing                                        1
Name: Activity, Length: 595, dtype: int64

In [27]:
pattern = '[Ss]wimming|[Dd]iving|[Ss]urfing|[Ff]ishing|[Ss]tanding|[Ww]alking|[Ff]ilming'
df_attacks['Activity'] = df_attacks['Activity'].apply(lambda x: re.findall(pattern, x)[0].lower() if len(re.findall(pattern, x))>0 else x.lower())
df_attacks['Activity'].value_counts()

surfing                                 1054
fishing                                  541
swimming                                 519
diving                                   293
unknown                                  219
                                        ... 
wading, when he stepped on the shark       1
attempting to retreive a dinghy            1
playing in the surf with his 2 dogs        1
collecting beche-de-mer                    1
sight-seeing                               1
Name: Activity, Length: 266, dtype: int64

In [28]:
valid_activities = ['surfing', 'swimming', 'diving', 'fishing', 'standing', 'paddling', 'wading', 'unknown', 'other']

In [29]:
df_attacks['Activity'] = df_attacks['Activity'].apply(lambda x: 'surfing' if 'surf' in x else x)
df_attacks['Activity'] = df_attacks['Activity'].apply(lambda x: 'paddling' if 'paddl' in x else x)
df_attacks['Activity'] = df_attacks['Activity'].apply(lambda x: 'swimming' if 'floating' in x else x)
df_attacks['Activity'] = df_attacks['Activity'].apply(lambda x: 'surfing' if 'boarding' in x else x)
df_attacks['Activity'] = df_attacks['Activity'].apply(lambda x: 'wading' if 'wading' in x else x)
df_attacks['Activity'].value_counts()

surfing                            1235
fishing                             541
swimming                            540
diving                              293
unknown                             219
                                   ... 
attempting to retreive a dinghy       1
collecting beche-de-mer               1
fell off banana boat                  1
hiking on the beach                   1
sight-seeing                          1
Name: Activity, Length: 204, dtype: int64

In [30]:
df_attacks['Activity'] = df_attacks['Activity'].apply(lambda x: x if x in valid_activities else 'other')

In [31]:
df_attacks['Activity'].value_counts()

surfing     1235
fishing      541
swimming     540
other        408
diving       293
unknown      219
wading       118
standing      67
paddling      24
Name: Activity, dtype: int64

In [32]:
df_attacks

Date    Year        Type    Country             Area  \
0     25-Jun-2018  2018.0     Boating        USA       California   
1     18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2     09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3     08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4     04-Jun-2018  2018.0    Provoked     MEXICO           Colima   
...           ...     ...         ...        ...              ...   
3441     Feb-1969  1969.0     Invalid  AUSTRALIA       Queensland   
3442  27-Jan-1969  1969.0  Unprovoked  AUSTRALIA  New South Wales   
3443     Jan-1969  1969.0  Unprovoked  AUSTRALIA         Victoria   
3444  Winter 1969  1969.0    Provoked    BAHAMAS        Eleuthera   
3888  16-Jul-1961  1971.0  Unprovoked     TURKEY         Anatolia   

                            Location  Activity Sex   Age  \
0        Oceanside, San Diego County  paddling    F   57   
1     St. Simon Island, Glynn County  standing    F   11   
2                       Habush, Oahu   surfing    M   48   
3                 Arrawarra Headland   surfing    M  NaN   
4                           La Ticla    diving    M  NaN   
...                              ...       ...  ...  ...   
3441                        Cooktown    diving    M   52   
3442                   Beecroft Head    diving    M   21   
3443                  Port MacDonnel   unknown    M  NaN   
3444                             NaN     other  NaN  NaN   
3888         ?nciralti Beach, ?zmir   swimming    M   16   

                                                 Injury Fatal (Y/N)  \
0     No injury to occupant, outrigger canoe and pad...           N   
1                            Minor injury to left thigh           N   
2          Injury to left lower leg from surfboard skeg           N   
3                             Minor injury to lower leg           N   
4     Lacerations to leg & hand shark PROVOKED INCIDENT           N   
...                                                 ...         ...   
3441                                         No details         NaN   
3442       Abrasions and lacerations to lower right leg           N   
3443                                      Minor injury            N   
3444  No injury to occupants, shark grabbed prop, Le...           N   
3888                                   Left leg injured           N   

               Time               Species   
0             18h00            White shark  
1     14h00  -15h00                    NaN  
2             07h45                    NaN  
3               NaN              2 m shark  
4               NaN        Tiger shark, 3m  
...             ...                    ...  
3441            NaN  Questionable incident  
3442          07h30                    NaN  
3443            NaN           Carpet shark  
3444            NaN              10' shark  
3888            NaN                    NaN  

[3445 rows x 13 columns]

---

### Cleaning Sex Column and renaming it to Gender

In [33]:
df_attacks['Sex '].value_counts()

M      2723
F       480
M         2
lli       1
Name: Sex , dtype: int64

This column also need to be treated.
1. The name of the column should be renamed;
2. Should have only one `M`;
3. If `lli` has no meaning it is better to be left as `Unknown`.

In [34]:
#renaming the column
df_attacks.rename(columns={'Sex ': 'Gender'}, inplace=True)
df_attacks.head(2)

Date    Year        Type Country        Area  \
0  25-Jun-2018  2018.0     Boating     USA  California   
1  18-Jun-2018  2018.0  Unprovoked     USA     Georgia   

                         Location  Activity Gender Age  \
0     Oceanside, San Diego County  paddling      F  57   
1  St. Simon Island, Glynn County  standing      F  11   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   

            Time     Species   
0          18h00  White shark  
1  14h00  -15h00          NaN

In [35]:
#checking if all the values are string type
df_attacks['Gender'].apply(type).value_counts()

<class 'str'>      3206
<class 'float'>     239
Name: Gender, dtype: int64

In [36]:
df_attacks['Gender'] = df_attacks['Gender'].apply(lambda x: x if type(x) == str else 'Unknown')
df_attacks['Gender'].apply(type).value_counts()

<class 'str'>    3445
Name: Gender, dtype: int64

In [37]:
df_attacks['Gender'] = df_attacks['Gender'].apply(lambda x: 'M' if 'M' in x else x)
df_attacks['Gender'] = df_attacks['Gender'].apply(lambda x: 'Unknown' if 'li' in x else x)
df_attacks['Gender'].value_counts()

M          2725
F           480
Unknown     240
Name: Gender, dtype: int64

---

### Cleaning Age

In [38]:
df_attacks['Age'].value_counts()

17          96
16          93
20          92
19          91
15          86
            ..
84           1
             1
30 or 36     1
6½           1
adult        1
Name: Age, Length: 118, dtype: int64

This column should be `int` or `float` type.

In [39]:
#checking the values type
df_attacks['Age'].apply(type).value_counts()

<class 'str'>      2453
<class 'float'>     992
Name: Age, dtype: int64

There is a lot of string type. To treat it mus be defined a function to try to convert the strings value into float, otherwise, it will defined as a null value.

In [40]:
#creating the function to convert the values
def convert_to_float (x):
    try:
        return float(x)
    except:
        return None

In [41]:
df_attacks['Age'] = df_attacks['Age'].apply(convert_to_float)
df_attacks['Age'].value_counts()

17.0    96
16.0    93
20.0    93
19.0    91
15.0    86
        ..
87.0     1
72.0     1
86.0     1
82.0     1
67.0     1
Name: Age, Length: 77, dtype: int64

In [42]:
#percentage of data that is unknown the age of the victim
df_attacks['Age'].isnull().sum()/len(df_attacks['Age'])

0.3053701015965167

---

In [43]:
df_attacks['Injury'].value_counts()

FATAL                                                                  242
Foot bitten                                                             64
Survived                                                                56
Left foot bitten                                                        46
Leg bitten                                                              43
                                                                      ... 
Drowned, body scavenged by shark                                         1
No injury to occupant, shark removed small auxiliary outboard motor      1
Severe lacerations to lower left leg                                     1
No injury, flung from surfboard by the shark                             1
Left leg injured                                                         1
Name: Injury, Length: 2110, dtype: int64

Too many `Injury` types. This must be simplified if it is to be used to answer the business question.

But I'm not going to use it for analysis, so I'll leave it as it is. probably will drop it further.

---

### Cleaning the Fatal (Y/N) column

In [44]:
df_attacks['Fatal (Y/N)'].value_counts()

N          2772
Y           383
UNKNOWN      29
M             1
2017          1
Name: Fatal (Y/N), dtype: int64

`M` probably `N` (typing error).

`2017` as `UNKNOWN`, unless the `Injury` gives a clue if it was fatal or not. This must be checked before converting non string values into string.

In [45]:
#checking the values type
df_attacks['Fatal (Y/N)'].apply(type).value_counts()

<class 'str'>      3186
<class 'float'>     259
Name: Fatal (Y/N), dtype: int64

In [46]:
df_attacks[df_attacks['Fatal (Y/N)'] == '2017']

Date    Year      Type Country      Area  Location Activity  \
786  10-Jun-2012  2012.0  Provoked   ITALY  Sardinia  Muravera    other   

    Gender   Age                                    Injury Fatal (Y/N)  \
786      M  57.0  Lower left leg injured PROVOKED ACCIDENT        2017   

        Time          Species   
786  Morning  Blue shark, 2.5m

It is not clear if the victim survived or not, so I'll treat it as `UNKNOWN`.
Notes that `2017` is a string type.

In [47]:
df_attacks['Fatal (Y/N)'] = df_attacks['Fatal (Y/N)'].apply(lambda x: x if type(x) == str else 'UNKNOWN')
df_attacks['Fatal (Y/N)'].apply(type).value_counts()

<class 'str'>    3445
Name: Fatal (Y/N), dtype: int64

In [48]:
df_attacks['Fatal (Y/N)'].value_counts()

N          2772
Y           383
UNKNOWN     288
M             1
2017          1
Name: Fatal (Y/N), dtype: int64

In [49]:
df_attacks['Fatal (Y/N)'] = df_attacks['Fatal (Y/N)'].apply(lambda x: 'N' if x == 'M' else x)
df_attacks['Fatal (Y/N)'] = df_attacks['Fatal (Y/N)'].apply(lambda x: x if x in 'YN' else 'UNKNOWN')
df_attacks['Fatal (Y/N)'].value_counts()

N          2773
Y           383
UNKNOWN     289
Name: Fatal (Y/N), dtype: int64

---

In [50]:
df_attacks['Time'].value_counts()

Afternoon              124
11h00                   93
Morning                 90
15h00                   80
16h00                   70
                      ... 
1300                     1
"Evening"                1
Just before sundown      1
19h05                    1
Daytime                  1
Name: Time, Length: 309, dtype: int64

I don't know how to deal with it. I'll probably not use this column...

---

### Cleaning Species column

In [51]:
df_attacks['Species '].value_counts()

White shark                                                                 121
Shark involvement not confirmed                                              69
Shark involvement prior to death was not confirmed                           56
Tiger shark                                                                  49
Invalid                                                                      43
                                                                           ... 
A 10-year-old 94-pound pregnant blacktip reef shark                           1
Lesser spotted dogfish, Scyliorhinus canicula, less than 80 cm in length      1
White shark, 2m                                                               1
Lemon shark, >1 m                                                             1
White shark, identified by tooth fragments in surfboard                       1
Name: Species , Length: 1056, dtype: int64

I'll probably not use this for analysis.
But very interesting statements poped up here: `Shark involvement not confirmed`, and `Shark involvement prior to death was not confirmed`. Theses statements makes its rows not reliable to count as a record of shark attack and must be dropped.

Deeper and further analysis should be done in this column to check if there is need for more row drops.

In [52]:
#renaming the column
df_attacks.rename(columns={'Species ': 'Species'}, inplace=True)

In [53]:
df_attacks['Species'].apply(type).value_counts()

<class 'str'>      2239
<class 'float'>    1206
Name: Species, dtype: int64

In [54]:
df_attacks['Species'] = df_attacks['Species'].apply(lambda x: x if type(x) == str else 'UNKNOWN')
df_attacks['Species'].apply(type).value_counts()

<class 'str'>    3445
Name: Species, dtype: int64

In [55]:
df_attacks['Species'].value_counts()

UNKNOWN                                                                     1206
White shark                                                                  121
Shark involvement not confirmed                                               69
Shark involvement prior to death was not confirmed                            56
Tiger shark                                                                   49
                                                                            ... 
A 10-year-old 94-pound pregnant blacktip reef shark                            1
Lesser spotted dogfish, Scyliorhinus canicula, less than 80 cm in length       1
White shark, 2m                                                                1
Lemon shark, >1 m                                                              1
White shark, identified by tooth fragments in surfboard                        1
Name: Species, Length: 1057, dtype: int64

In [56]:
df_attacks['Species'][df_attacks['Species'].str.contains(('confirmed') and ('death'))].value_counts()

Shark involvement prior to death was not confirmed                                           56
Shark involvement prior to death not confirmed                                                9
Shark involvement prior to death suspected but not confirmed                                  2
Shark involvement prior to death still to be determined                                       1
Reported by media as shark attack, but shark involvement prior to death was not confirmed     1
Shark involvement prior to death unconfired                                                   1
Shark involvement prior to death could not be determined                                      1
Shark involvement prior to death remains unconfirmed                                          1
Name: Species, dtype: int64

In [57]:
df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement prior to death was not confirmed'\
 if x == 'Shark involvement prior to death not confirmed' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement prior to death was not confirmed'\
 if x == 'Shark involvement prior to death suspected but not confirmed' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement prior to death was not confirmed'\
 if x == 'Shark involvement prior to death still to be determined' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement prior to death was not confirmed'\
 if x == 'Reported by media as shark attack, but shark involvement prior to death was not confirmed' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement prior to death was not confirmed'\
 if x == 'Shark involvement prior to death unconfired' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement prior to death was not confirmed'\
 if x == 'Shark involvement prior to death could not be determined' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement prior to death was not confirmed'\
 if x == 'Shark involvement prior to death remains unconfirmed' else x)

df_attacks['Species'][df_attacks['Species'].str.contains(('confirmed') and ('death'))].value_counts()

Shark involvement prior to death was not confirmed    72
Name: Species, dtype: int64

In [58]:
df_attacks['Species'].value_counts()

UNKNOWN                                                                     1206
White shark                                                                  121
Shark involvement prior to death was not confirmed                            72
Shark involvement not confirmed                                               69
Tiger shark                                                                   49
                                                                            ... 
A 10-year-old 94-pound pregnant blacktip reef shark                            1
Lesser spotted dogfish, Scyliorhinus canicula, less than 80 cm in length       1
White shark, 2m                                                                1
Lemon shark, >1 m                                                              1
White shark, identified by tooth fragments in surfboard                        1
Name: Species, Length: 1050, dtype: int64

In [59]:
df_attacks['Species'][df_attacks['Species'].str.contains(('confirmed'))].value_counts()

Shark involvement prior to death was not confirmed                                                 72
Shark involvement not confirmed                                                                    69
Shark involvement suspected but not confirmed                                                       3
Shark involvement unconfirmed                                                                       1
White shark, 3.5 m [11.5'], species identity confirmed by tooth fragment                            1
Said to involve a white shark but shark involvement not confirmed                                   1
2.4 m [8'] white shark, species identity confirmed by witnesses & tooth pattern in surfboard        1
2.4 m [8'] white shark, species identity confirmed by tooth fragment                                1
Shark involvement not confirmed; officials considered barracua                                      1
Doubtful / Unconfirmed attack / Unable to verify in local records                 

In [60]:
df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement suspected but not confirmed' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement unconfirmed' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Said to involve a white shark but shark involvement not confirmed' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement not confirmed; officials considered barracua' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Doubtful / Unconfirmed attack / Unable to verify in local records' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement  not confirmed' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement not confirmed, injury may be due to a stingray' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement not confirmed, injury may have been caused by a bluefish' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement not confirmed; thought to be a barracuda bite' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement probable, but not confirmed' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'shark involvement not confirmed' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement not confirmed & highly unlikely' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == "Thought to involve a 3' to 4' shark, but shark involvement not confirmed" else x)

df_attacks['Species'][df_attacks['Species'].str.contains(('confirmed'))].value_counts()

Shark involvement not confirmed                                                                    84
Shark involvement prior to death was not confirmed                                                 72
2.4 m [8'] white shark, species identity confirmed by tooth fragment                                1
2.4 m [8'] white shark, species identity confirmed by witnesses & tooth pattern in surfboard        1
White shark, 3.5 m [11.5'], species identity confirmed by tooth fragment                            1
White shark, 3 m [10'], species identity confirmed by witnesses & tooth pattern in leg & board      1
Name: Species, dtype: int64

In [61]:
df_attacks['Species'][df_attacks['Species'].str.contains(('involvement'))].value_counts()

Shark involvement not confirmed                                                                                                                                                            84
Shark involvement prior to death was not confirmed                                                                                                                                         72
No shark involvement                                                                                                                                                                       21
Shark involvement questionable                                                                                                                                                              3
Shark involvement not cofirmed                                                                                                                                                              2
Shark involvement doubtful                        

In [62]:
df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'No shark involvement' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement questionable' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement questionable' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement not cofirmed' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement doubtful' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == 'Shark involvement highly doubtful' else x)

df_attacks['Species'] = df_attacks['Species'].apply\
(lambda x: 'Shark involvement not confirmed'\
 if x == "Mr. Burgess of ISAF announced the injury was the bite of a 1.8 m [6'], 2- to 3-year old white shark. Subsequent investigation revealed there was  no shark involvement in this incident" else x)

df_attacks['Species'][df_attacks['Species'].str.contains(('confirmed'))].value_counts()

Shark involvement not confirmed                                                                    114
Shark involvement prior to death was not confirmed                                                  72
2.4 m [8'] white shark, species identity confirmed by tooth fragment                                 1
2.4 m [8'] white shark, species identity confirmed by witnesses & tooth pattern in surfboard         1
White shark, 3.5 m [11.5'], species identity confirmed by tooth fragment                             1
White shark, 3 m [10'], species identity confirmed by witnesses & tooth pattern in leg & board       1
Name: Species, dtype: int64

Need to check if thoses `Shark involvement prior to death was not confirmed` are recorded as fatal attack.

In [63]:
df_attacks['Fatal (Y/N)'][df_attacks['Species'] == 'Shark involvement prior to death was not confirmed'].value_counts()

UNKNOWN    68
Y           3
N           1
Name: Fatal (Y/N), dtype: int64

In [64]:
df_attacks[(df_attacks['Type'] != 'Invalid') & (df_attacks['Species'] == 'Shark involvement not confirmed')]

Date    Year          Type    Country             Area  \
16    09-May-2018  2018.0  Questionable  AUSTRALIA  New South Wales   
294   28-Mar-2016  2016.0    Unprovoked        USA          Florida   
557   11-May-2014  2014.0    Unprovoked        USA          Georgia   
782   14-Jun-2012  2012.0    Unprovoked        USA   South Carolina   
1285  15-Mar-2008  2008.0    Unprovoked        USA          Florida   
2593    July 1991  1991.0    Unprovoked        USA         Virginia   
2854  26-May-1985  1985.0    Unprovoked        USA       California   

                                               Location  Activity Gender  \
16                               Sharpes Beach, Ballina   surfing      M   
294                        Fort Myers Beach, Lee County   unknown      M   
557                        Tybee Island, Chatham County   surfing      M   
782                          Myrtle Beach, Horry County  swimming      M   
1285  Lovers Key State Park, Bonita Springs, Lee County     other      M   
2593                    Croatan Beach, Virginia Beach,    surfing      M   
2854                     Long Beach, Los Angeles County   surfing      M   

       Age                                             Injury Fatal (Y/N)  \
16     NaN                       No injury, surfboard damaged           N   
294    NaN  Minor injury to arm. Possibly caused by smalll...     UNKNOWN   
557   12.0                     Puncture wounds to right thigh           N   
782    NaN                                       Minor injury           N   
1285   8.0                                       Minor injury           N   
2593  13.0                          Severe laceration to foot           N   
2854   NaN                                        Leg injured           N   

           Time                          Species  
16        10h30  Shark involvement not confirmed  
294         NaN  Shark involvement not confirmed  
557         NaN  Shark involvement not confirmed  
782   Afternoon  Shark involvement not confirmed  
1285  Afternoon  Shark involvement not confirmed  
2593        NaN  Shark involvement not confirmed  
2854        NaN  Shark involvement not confirmed

In [65]:
df_attacks['Fatal (Y/N)'][df_attacks['Species'] == 'Shark involvement not confirmed'].value_counts()

UNKNOWN    104
N           10
Name: Fatal (Y/N), dtype: int64

Most of the "not confirmed shark involvement" prior to death or not is reported as `UNKNOWN` if it was fatal or not.
,

---

### Droping rows of records that are not confirmed as a shark attack

Since is known that 114 observations were not confirmed as a shark attack, and other 72 were not confirmed as shark attack prior to death, those observations should be excluded from tha analysis.

In [66]:
df_attacks.drop(df_attacks[df_attacks['Species'] == 'Shark involvement prior to death was not confirmed'].index, inplace=True)
df_attacks.drop(df_attacks[df_attacks['Species'] == 'Shark involvement not confirmed'].index, inplace=True)

In [67]:
df_attacks.shape

(3259, 13)

#### Deeper analysis on the Type column

In [68]:
df_attacks.Type.value_counts()

Unprovoked      2691
Provoked         257
Boating          162
Invalid           95
Sea Disaster      51
Unknown            2
Questionable       1
Name: Type, dtype: int64

There is still 95 `Invalid` and 1 `Questionable` and 2 `Unknown`. Are thoses records really of a shark attack?

In [69]:
df_attacks[df_attacks['Type'] == 'Questionable']

Date    Year          Type    Country             Area  \
20  25-Apr-2018  2018.0  Questionable  AUSTRALIA  New South Wales   

       Location Activity Gender  Age     Injury Fatal (Y/N)   Time  \
20  Lennox Head  surfing      M  NaN  No injury           N  07h00   

         Species  
20  Questionable

In [70]:
df_attacks['Species'][df_attacks['Type'] == 'Invalid'].value_counts()

Invalid                                                                                       43
UNKNOWN                                                                                       13
Questionable incident                                                                         13
Questionable Incident                                                                          3
White shark                                                                                    2
Tiger shark, 13' female                                                                        1
White shark?                                                                                   1
Reported as a shark attack, the story was a hoax                                               1
Questionable incident; reported as shark attack but thought to involve a pinniped instead      1
Shark involvement  questionable                                                                1
Questionable incident - shark 

When crossing the information in the `Species` column when `Type` is invalid, it shows that most of them are questionable to be considered as shark attack, but it also shows some excpetions.
The `Type` of those exceptions will be treated as `Valid, but unknown`.

In [71]:
df_attacks['Type'] = df_attacks.apply\
(lambda x: 'Valid, but unknown'\
 if x['Species'] == 'White shark' else x['Type'], axis=1)

df_attacks['Type'] = df_attacks.apply\
(lambda x: 'Valid, but unknown'\
 if x['Species'] == "Tiger shark, 13' female" else x['Type'], axis=1)

df_attacks['Type'] = df_attacks.apply\
(lambda x: 'Valid, but unknown'\
 if x['Species'] == 'White shark?' else x['Type'], axis=1)

df_attacks['Type'] = df_attacks.apply\
(lambda x: 'Valid, but unknown'\
 if x['Species'] == '24" to 30" shark' else x['Type'], axis=1)

df_attacks['Type'] = df_attacks.apply\
(lambda x: 'Valid, but unknown'\
 if x['Species'] == 'A small shark' else x['Type'], axis=1)

df_attacks['Type'] = df_attacks.apply\
(lambda x: 'Valid, but unknown'\
 if x['Species'] == '2m shark' else x['Type'], axis=1)

In [72]:
df_attacks['Type'].value_counts()

Unprovoked            2605
Provoked               255
Valid, but unknown     130
Boating                127
Invalid                 88
Sea Disaster            51
Unknown                  2
Questionable             1
Name: Type, dtype: int64

In [73]:
df_attacks[df_attacks['Type'] == "Unknown"]

Date    Year     Type    Country           Area          Location  \
85   15-Sep-2017  2017.0  Unknown      SAMOA   Upolu Island         Nofoali’i   
382  27-Jul-2015  2015.0  Unknown  AUSTRALIA       Victoria  Tyrendarra Beach   

    Activity Gender   Age                      Injury Fatal (Y/N)   Time  \
85   fishing      M   NaN  Injuries to hands and legs           N  Night   
382  surfing      M  40.0              Injury to hand     UNKNOWN    NaN   

     Species  
85   UNKNOWN  
382  UNKNOWN

In [74]:
(df_attacks['Type'][df_attacks['Type'] == "Unknown"].count()\
 +df_attacks['Type'][df_attacks['Type'] == "Invalid"].count()\
 +df_attacks['Type'][df_attacks['Type'] == "Questionable"].count())\
/len(df_attacks)

0.027922675667382633

As it cannot be determined if the records Invalid, Questinable and Unknown in `Type` column were a shark attack or not, and it represents only 2.8% of the dataset, all those rows will be dropped.

In [75]:
df_attacks.drop(df_attacks[df_attacks['Type'] == 'Invalid'].index, inplace=True)
df_attacks.drop(df_attacks[df_attacks['Type'] == 'Unknown'].index, inplace=True)
df_attacks.drop(df_attacks[df_attacks['Type'] == 'Questionable'].index, inplace=True)

In [76]:
df_attacks.shape

(3168, 13)

---

### Dropping columns that will not be used on the analysis

The columns `Date`, `Injury` and `Time` will not be used for the further analysis, so it will be dropped.
The column `Species` will not be used as well, but it was useful during the data cleaning, and will be dropped now.
The columns `Country`, `Area` and `Location` were not clean yet. Thoses I will keep, because I think it may be useful in the future.

In [77]:
df_attacks.drop(columns=['Date', 'Injury', 'Time', 'Species'], inplace=True)

In [78]:
df_attacks

Year                Type    Country             Area  \
0     2018.0  Valid, but unknown        USA       California   
1     2018.0          Unprovoked        USA          Georgia   
3     2018.0          Unprovoked  AUSTRALIA  New South Wales   
4     2018.0            Provoked     MEXICO           Colima   
5     2018.0          Unprovoked  AUSTRALIA  New South Wales   
...      ...                 ...        ...              ...   
3440  1969.0            Provoked  AUSTRALIA         Tasmania   
3442  1969.0          Unprovoked  AUSTRALIA  New South Wales   
3443  1969.0          Unprovoked  AUSTRALIA         Victoria   
3444  1969.0            Provoked    BAHAMAS        Eleuthera   
3888  1971.0          Unprovoked     TURKEY         Anatolia   

                            Location  Activity   Gender   Age Fatal (Y/N)  
0        Oceanside, San Diego County  paddling        F  57.0           N  
1     St. Simon Island, Glynn County  standing        F  11.0           N  
3                 Arrawarra Headland   surfing        M   NaN           N  
4                           La Ticla    diving        M   NaN           N  
5                 Flat Rock, Ballina   surfing        M   NaN           N  
...                              ...       ...      ...   ...         ...  
3440                         Taroona   fishing        M  49.0           N  
3442                   Beecroft Head    diving        M  21.0           N  
3443                  Port MacDonnel   unknown        M   NaN           N  
3444                             NaN     other  Unknown   NaN           N  
3888         ?nciralti Beach, ?zmir   swimming        M  16.0           N  

[3168 rows x 9 columns]

---

### Saving to a new .csv the cleaned dataset

In [79]:
#saving new clean .csv
df_attacks.to_csv('data/attacks_cleaned.csv', index=False)

In [80]:
#testing the new file
pd.read_csv('data/attacks_cleaned.csv')

Year                Type    Country             Area  \
0     2018.0  Valid, but unknown        USA       California   
1     2018.0          Unprovoked        USA          Georgia   
2     2018.0          Unprovoked  AUSTRALIA  New South Wales   
3     2018.0            Provoked     MEXICO           Colima   
4     2018.0          Unprovoked  AUSTRALIA  New South Wales   
...      ...                 ...        ...              ...   
3163  1969.0            Provoked  AUSTRALIA         Tasmania   
3164  1969.0          Unprovoked  AUSTRALIA  New South Wales   
3165  1969.0          Unprovoked  AUSTRALIA         Victoria   
3166  1969.0            Provoked    BAHAMAS        Eleuthera   
3167  1971.0          Unprovoked     TURKEY         Anatolia   

                            Location  Activity   Gender   Age Fatal (Y/N)  
0        Oceanside, San Diego County  paddling        F  57.0           N  
1     St. Simon Island, Glynn County  standing        F  11.0           N  
2                 Arrawarra Headland   surfing        M   NaN           N  
3                           La Ticla    diving        M   NaN           N  
4                 Flat Rock, Ballina   surfing        M   NaN           N  
...                              ...       ...      ...   ...         ...  
3163                         Taroona   fishing        M  49.0           N  
3164                   Beecroft Head    diving        M  21.0           N  
3165                  Port MacDonnel   unknown        M   NaN           N  
3166                             NaN     other  Unknown   NaN           N  
3167         ?nciralti Beach, ?zmir   swimming        M  16.0           N  

[3168 rows x 9 columns]